File: tests/lua-empty-test/src/DragonBonesHandler.lua
require('utils')
require('DragonBoneBullet')

local NORMAL_ANIMATION_GROUP = 'normal'
local AIM_ANIMATION_GROUP = 'aim'
local ATTACK_ANIMATION_GROUP = 'attack'
local JUMP_SPEED = -20
local NORMALIZE_MOVE_SPEED = 3.6
local MAX_MOVE_SPEED_FRONT = NORMALIZE_MOVE_SPEED * 1.4
local MAX_MOVE_SPEED_BACK = NORMALIZE_MOVE_SPEED * 1

local WEAPON_R_LIST = {'weapon_1502b_r', 'weapon_1005', 'weapon_1005b', 'weapon_1005c', 'weapon_1005d', 'weapon_1005e'}
local WEAPON_L_LIST = {'weapon_1502b_l', 'weapon_1005', 'weapon_1005b', 'weapon_1005c', 'weapon_1005d'}

local GROUND = 120
local G = -0.6

local TOUCH_EVENT_BEGAN = 0
local TOUCH_EVENT_ENDED = 2

local function generateAnimationEventHandler(target)
    return function(event)
               target:animationEventHandler(event)
           end
end

local function generateFrameEventHandler(target)
    return function(event)
               target:frameEventHandler(event)
           end
end

DragonBonesHandler = class("DragonBonesHandler", function ()
    local node = cc.Node:create()
    return node
end)

function DragonBonesHandler:ctor(scene)
    -- body
    self._scene = scene
    self._left = false
    self._right = false
    self._isJumpingA = false
    self._isJumpingB = false
    self._isAttackingA = false
    self._isAttackingB = false
    self._isSquating = false
    self._weaponRIndex = 1
    self._weaponLIndex = 1
    self._faceDir = 1
    self._aimDir = 0
    self._moveDir = 0
    self._animRadian = 0
    self._speedX = 0
    self._speedY = 0
    self._animState = nil
    self._walkState = nil
    self._attackState = nil
    self._target = nil
    self._bullets = {}

    self._armatureDisplay = findChild(scene, 'DragonBoneNode', 'ccdb.CCArmatureDisplay')
    self._armature = self._armatureDisplay:getArmature()

    self._armatureDisplay:addEvent(ccdb.EventObject.FADE_IN_COMPLETE, generateAnimationEventHandler(self))
    self._armatureDisplay:addEvent(ccdb.EventObject.FADE_OUT_COMPLETE, generateAnimationEventHandler(self))

    self._armature:getSlot('effects_1').displayController = NORMAL_ANIMATION_GROUP
    self._armature:getSlot('effects_2').displayController = NORMAL_ANIMATION_GROUP

    self._weaponR = self._armature:getSlot('weapon_r'):getChildArmature()
    self._weaponL = self._armature:getSlot('weapon_l'):getChildArmature()

    self._weaponR:getDisplay():addEvent(ccdb.EventObject.FRAME_EVENT, generateFrameEventHandler(self))
    self._weaponL:getDisplay():addEvent(ccdb.EventObject.FRAME_EVENT, generateFrameEventHandler(self))

    self:updateAnimation()

    -- up
    local button = findChild(scene, 'up', 'ccui.Button')
    button:addClickEventListener(function ()
    	self:jump()
    end)

    -- down
    button = findChild(scene, 'down', 'ccui.Button')
    button:addTouchEventListener(function (sender, type)
    	if type == TOUCH_EVENT_BEGAN then
    		self:squat(true)
    	elseif type == TOUCH_EVENT_ENDED then
    		self:squat(false)
    	end
    end)

    -- left
    button = findChild(scene, 'left', 'ccui.Button')
    button:addTouchEventListener(function (sender, type)
    	if type == TOUCH_EVENT_BEGAN then
    		self._left = true
    	elseif type == TOUCH_EVENT_ENDED then
    		self._left = false
    	end

    	self:updateMove(-1)
    end)

    -- right
    button = findChild(scene, 'right', 'ccui.Button')
    button:addTouchEventListener(function (sender, type)
    	if type == TOUCH_EVENT_BEGAN then
    		self._right = true
    	elseif type == TOUCH_EVENT_ENDED then
    		self._right = false
    	end

    	self:updateMove(-1)
    end)

    -- change weapon L
    button = findChild(scene, 'ChangeWeaponL', 'ccui.Button')
    button:addClickEventListener(function ()
    	self:switchWeaponL()
    end)

    -- change weapon R
    button = findChild(scene, 'ChangeWeaponR', 'ccui.Button')
    button:addClickEventListener(function ()
    	self:switchWeaponR()
    end)

    local scheduler = cc.Director:getInstance():getScheduler()
    self.scheduleEntry = scheduler:scheduleScriptFunc(function (dt)
        self:updatePosition()
        self:updateAnim()
        self:updateAttack()
        self:enterFrameHandler(dt)
    end, 0, false)

    self:registerScriptHandler(function(tag)
        if "enter" == tag then
            self:onEnter()
        elseif "exit" == tag then
        	self:onExit()
        end
    end)
end

function DragonBonesHandler:onEnter()

    function onTouchBegan(touches, event)
        self:attack(true)
        return true
    end

    function onTouchMoved(touches, event)
        local touchloc = touches[1]:getLocation()
        self:aim(touchloc.x, touchloc.y)
    end

    function onTouchEnded(touches, event)
        self:attack(false)
    end

	local listener = cc.EventListenerTouchAllAtOnce:create()
    listener:registerScriptHandler(onTouchBegan, cc.Handler.EVENT_TOUCHES_BEGAN)
    listener:registerScriptHandler(onTouchMoved, cc.Handler.EVENT_TOUCHES_MOVED)
    listener:registerScriptHandler(onTouchEnded, cc.Handler.EVENT_TOUCHES_ENDED)

    local eventDispatcher = cc.Director:getInstance():getEventDispatcher()
    eventDispatcher:addEventListenerWithSceneGraphPriority(listener, self._armatureDisplay) 
end

function DragonBonesHandler:onExit()
	cc.Director:getInstance():getEventDispatcher():removeEventListenersForTarget(self._armatureDisplay)

    local scheduler = cc.Director:getInstance():getScheduler()
    scheduler:unscheduleScriptEntry(self.scheduleEntry)
end

function DragonBonesHandler:animationEventHandler(event)
	if event.type == ccdb.EventObject.FADE_IN_COMPLETE then
		if event.animationState:getName() == 'jump_1' then
			self._isJumpingB = true
			self._speedY = self._speedY - JUMP_SPEED
			self._armature:getAnimation():fadeIn('jump_2', -1, -1, 0, NORMAL_ANIMATION_GROUP)
		elseif event.animationState:getName() == 'jump_4' then
			self:updateAnimation()
		end
	elseif event.type == ccdb.EventObject.FADE_OUT_COMPLETE then
        if event.animationState:getName() == 'attack_01' then
            self._isAttackingB = false
            self._attackState = nil
        end
	end
end

function DragonBonesHandler:updateAnimation()
	if self._isJumpingA then
        return
    end

    if self._isSquating then
        self._speedX = 0
        self._armature:getAnimation():fadeIn('squat', -1, -1, 0, NORMAL_ANIMATION_GROUP)
        self._walkState = nil
        return;
    end

    if self._moveDir == 0 then
        self._speedX = 0
        self._armature:getAnimation():fadeIn('idle', -1, -1, 0, NORMAL_ANIMATION_GROUP)
        self._walkState = nil
    else
        if not self._walkState then
            self._walkState = self._armature:getAnimation():fadeIn('walk', -1, -1, 0, NORMAL_ANIMATION_GROUP)
        end

        if self._moveDir * self._faceDir > 0 then
            self._walkState.timeScale = MAX_MOVE_SPEED_FRONT / NORMALIZE_MOVE_SPEED
            self._speedX = MAX_MOVE_SPEED_FRONT * self._faceDir
        else
            self._walkState.timeScale = -MAX_MOVE_SPEED_BACK / NORMALIZE_MOVE_SPEED
            self._speedX = -MAX_MOVE_SPEED_BACK * self._faceDir
        end
    end
end

function DragonBonesHandler:frameEventHandler(event)
	if event.animationState:getName() == 'fire_01' then
        local firePointBone = event.armature:getBone('firePoint')
        local localPoint = cc.p(firePointBone.global.x, -firePointBone.global.y)

        local display = event.armature:getDisplay()
        local globalPoint = display:convertToWorldSpace(localPoint)

        self:fire(globalPoint)
    end
end

function DragonBonesHandler:jump()
	if self._isJumpingA then
        return
    end

    self._isJumpingA = true
    self._armature:getAnimation():fadeIn('jump_1', -1, -1, 0, NORMAL_ANIMATION_GROUP)
    self._walkState = nil
end

function DragonBonesHandler:squat(isSquating)
	if isSquating == self._isSquating then
        return
    end

    self._isSquating = isSquating
    self:updateAnimation()
end

function DragonBonesHandler:updateMove(dir)
	if self._left and self._right then
        self:move(dir)
    elseif self._left then
        self:move(-1)
    elseif self._right then
        self:move(1)
    else
        self:move(0)
    end
end

function DragonBonesHandler:move(dir)
	if self._moveDir == dir then
        return
    end

    self._moveDir = dir
    self:updateAnimation()
end

function DragonBonesHandler:aim(x, y)
	if self._aimDir == 0 then
        self._aimDir = 10
    end

    self._target = self._armatureDisplay:getParent():convertToNodeSpace(cc.p(x, y))
end

function DragonBonesHandler:attack(isAttacking)
	if self._isAttackingA == isAttacking then
        return
    end

    self._isAttackingA = isAttacking
end

function DragonBonesHandler:updatePosition()
	if self._speedX ~= 0 then
        local armatureDisplayPosX, _ = self._armatureDisplay:getPosition()
        self._armatureDisplay:setPositionX(armatureDisplayPosX + self._speedX)
        local visibleRect = cc.Director:getInstance():getVisibleSize()
        local visibleOrigin = cc.Director:getInstance():getVisibleOrigin()
        local minX = visibleOrigin.x + 10
        local maxX = visibleRect.width + visibleOrigin.x
        if armatureDisplayPosX < minX then
            self._armatureDisplay:setPositionX(minx)
        elseif armatureDisplayPosX > maxX then
            self._armatureDisplay:setPositionX(maxX)
        end
    end

    if self._speedY ~= 0 then
        if self._speedY > 5 and (self._speedY + G) <= 5 then
            self._armature:getAnimation():fadeIn('jump_3', -1, -1, 0, NORMAL_ANIMATION_GROUP)
        end

        self._speedY = self._speedY + G;

        local _, armatureDisplayPosY = self._armatureDisplay:getPosition()
        self._armatureDisplay:setPositionY(armatureDisplayPosY + self._speedY)
        if armatureDisplayPosY < GROUND then
            self._armatureDisplay:setPositionY(GROUND)
            self._isJumpingA = false
            self._isJumpingB = false
            self._speedX = 0
            self._speedY = 0
            self._armature:getAnimation():fadeIn('jump_4', -1, -1, 0, NORMAL_ANIMATION_GROUP)
            if self._isSquating or self._moveDir then
                self:updateAnimation()
            end
        end
    end
end

function DragonBonesHandler:updateAnim()
	if self._aimDir == 0 then
        return
    end

    local x, y = self._armatureDisplay:getPosition()
    if self._target.x > x then
        self._faceDir = 1
    else
        self._faceDir = -1
    end
    if self._armatureDisplay:getScaleX() * self._faceDir < 0 then
        self._armatureDisplay:setScaleX(self._armatureDisplay:getScaleX() * -1)
        if self._moveDir then
            self:updateAnimation()
        end
    end

    local animOffsetY = self._armature:getBone('chest').global.y * self._armatureDisplay:getScaleY();
    if self._faceDir > 0 then
        self._animRadian = math.atan2(-(self._target.y - y + animOffsetY), self._target.x - x)
    else
        self._animRadian = math.pi - math.atan2(-(self._target.y - y + animOffsetY), self._target.x - x)
        if self._animRadian > math.pi then
            self._animRadian = self._animRadian - math.pi * 2
        end
    end

    local aimDir = 0
    if self._animRadian > 0 then
        aimDir = -1
    else
        aimDir = 1
    end

    if self._aimDir ~= aimDir then
        self._aimDir = aimDir

        if self._aimDir >= 0 then
            self._animState = self._armature:getAnimation():fadeIn('aimUp', 0, 1, 0, AIM_ANIMATION_GROUP, ccdb.AnimationFadeOutMode.SameGroup)
        else
            self._animState = self._armature:getAnimation():fadeIn('aimDown', 0, 1, 0, AIM_ANIMATION_GROUP, ccdb.AnimationFadeOutMode.SameGroup)
        end
    end

    self._animState.weight = math.abs(self._animRadian / math.pi * 2)
    self._armature:invalidUpdate()
end

function DragonBonesHandler:updateAttack()
	if not self._isAttackingA or self._isAttackingB then
        return
    end

    self._isAttackingB = true

    self._attackState = self._armature:getAnimation():fadeIn('attack_01', -1, -1, 0, ATTACK_ANIMATION_GROUP, ccdb.AnimationFadeOutMode.SameGroup)

    self._attackState.autoFadeOutTime = self._attackState.fadeTotalTime;
    self._attackState:addBoneMask('pelvis')
end

function DragonBonesHandler:enterFrameHandler(dt)
	for i = #self._bullets, 1, -1 do
        local bullet = self._bullets[i]
        if bullet:update() then
            table.remove(self._bullets, i)
        end
    end

    ccdb.WorldClock.clock:advanceTime(dt)
end

function DragonBonesHandler:switchWeaponL()
	self._weaponL:getDisplay():removeEvent(ccdb.EventObject.FRAME_EVENT)

    self._weaponLIndex = (self._weaponLIndex + 1) % (#WEAPON_L_LIST + 1)
    if self._weaponLIndex == 0 then
        self._weaponLIndex = 1
    end
    local newWeaponName = WEAPON_L_LIST[self._weaponLIndex]
    self._weaponL = ccdb.CCFactory.factory:buildArmature(newWeaponName)
    self._armature:getSlot('weapon_l'):setChildArmature(self._weaponL)

    self._weaponL:getDisplay():addEvent(ccdb.EventObject.FRAME_EVENT, generateFrameEventHandler(self))
end

function DragonBonesHandler:switchWeaponR()
    self._weaponR:getDisplay():removeEvent(ccdb.EventObject.FRAME_EVENT)

    self._weaponRIndex = (self._weaponRIndex + 1) % (#WEAPON_R_LIST + 1)
    if self._weaponRIndex == 0 then
        self._weaponRIndex = 1
    end
    local newWeaponName = WEAPON_R_LIST[self._weaponRIndex]
    self._weaponR = ccdb.CCFactory.factory:buildArmature(newWeaponName)
    self._armature:getSlot('weapon_r'):setChildArmature(self._weaponR)

    self._weaponR:getDisplay():addEvent(ccdb.EventObject.FRAME_EVENT, generateFrameEventHandler(self))
end

function DragonBonesHandler:fire(pos)
	local firePoint = pos
    firePoint.x = firePoint.x + math.random() * 2 - 1
    firePoint.y = firePoint.y + math.random() * 2 - 1

    local armature = ccdb.CCFactory.factory:buildArmature('bullet_01')
    local effect = ccdb.CCFactory.factory:buildArmature('fireEffect_01')

    local radian = nil
    if self._faceDir < 0 then
        radian = math.pi - self._animRadian
    else
        radian = self._animRadian
    end

    local bullet = DragonBoneBullet.new(self:getParent(), armature, effect, radian + math.random() * 0.02 - 0.01, 40, firePoint)
    self:addBullet(bullet)
end

function DragonBonesHandler:addBullet(bullet)
	table.insert(self._bullets, bullet)
end

File: tests/cpp-empty-test/Classes/DragonBonesHandler.cpp
#include "DragonBonesHandler.h"

#include "ui/UIButton.h"

#include "dragonbones/DragonBonesHeaders.h"
#include "dragonbones/cocos2dx/CCDragonBonesHeaders.h"


std::string DragonBonesHandler::NORMAL_ANIMATION_GROUP = "normal";
std::string DragonBonesHandler::AIM_ANIMATION_GROUP = "aim";
std::string DragonBonesHandler::ATTACK_ANIMATION_GROUP = "attack";
float DragonBonesHandler::JUMP_SPEED = -20.f;
float DragonBonesHandler::NORMALIZE_MOVE_SPEED = 3.6f;
float DragonBonesHandler::MAX_MOVE_SPEED_FRONT = DragonBonesHandler::NORMALIZE_MOVE_SPEED * 1.4f;
float DragonBonesHandler::MAX_MOVE_SPEED_BACK = DragonBonesHandler::NORMALIZE_MOVE_SPEED * 1.f;
std::vector<std::string> DragonBonesHandler::WEAPON_R_LIST = {
    "weapon_1502b_r", "weapon_1005", "weapon_1005b", "weapon_1005c", "weapon_1005d", "weapon_1005e"
};
std::vector<std::string> DragonBonesHandler::WEAPON_L_LIST = {
    "weapon_1502b_l", "weapon_1005", "weapon_1005b", "weapon_1005c", "weapon_1005d"
};
float DragonBonesHandler::GROUND = 120;
float DragonBonesHandler::G = -0.6f;

using namespace dragonBones;

DragonBonesHandler::DragonBonesHandler(cocos2d::Scene* scene)
: _scene(scene)
, _left(false)
, _right(false)
, _isJumpingA(false)
, _isJumpingB(false)
, _isAttackingA(false)
, _isAttackingB(false)
, _isSquating(false)
, _weaponRIndex(0)
, _weaponLIndex(0)
, _faceDir(1)
, _aimDir(0)
, _moveDir(0)
, _animRadian(0.f)
, _speedX(0.f)
, _speedY(0.f)
, _animState(nullptr)
, _walkState(nullptr)
, _attackState(nullptr)
, _target(0, 0)
{
    _armatureDisplay = cocos2d::utils::findChild<CCArmatureDisplay*>(_scene, "DragonBoneNode");
    _armature = _armatureDisplay->_armature;
    
    
    _armatureDisplay->addEvent(EventObject::FADE_IN_COMPLETE, CC_CALLBACK_1(DragonBonesHandler::animationEventHandler, this));
    _armatureDisplay->addEvent(EventObject::FADE_OUT_COMPLETE, CC_CALLBACK_1(DragonBonesHandler::animationEventHandler, this));
    
    
    _armature->getSlot("effects_1")->displayController = DragonBonesHandler::NORMAL_ANIMATION_GROUP;
    _armature->getSlot("effects_2")->displayController = DragonBonesHandler::NORMAL_ANIMATION_GROUP;
    
    _weaponR = this->_armature->getSlot("weapon_r")->getChildArmature();
    _weaponL = this->_armature->getSlot("weapon_l")->getChildArmature();
    static_cast<CCArmatureDisplay*>(_weaponR->_display)->addEvent(EventObject::FRAME_EVENT, CC_CALLBACK_1(DragonBonesHandler::frameEventHandler, this));
    static_cast<CCArmatureDisplay*>(_weaponL->_display)->addEvent(EventObject::FRAME_EVENT, CC_CALLBACK_1(DragonBonesHandler::frameEventHandler, this));
    
    updateAnimation();
    
    // up
    auto button = cocos2d::utils::findChild<cocos2d::ui::Button*>(_scene, "up");
    button->addClickEventListener([=](cocos2d::Ref*){
        this->jump();
    });
    
    // down
    button = cocos2d::utils::findChild<cocos2d::ui::Button*>(_scene, "down");
    button->addTouchEventListener([=](cocos2d::Ref*, cocos2d::ui::Widget::TouchEventType type){
        if (type == cocos2d::ui::Widget::TouchEventType::BEGAN)
            this->squat(true);
        if (type == cocos2d::ui::Widget::TouchEventType::ENDED)
            this->squat(false);
    });
    
    // left
    button = cocos2d::utils::findChild<cocos2d::ui::Button*>(_scene, "left");
    button->addTouchEventListener([=](cocos2d::Ref*, cocos2d::ui::Widget::TouchEventType type){
        if (type == cocos2d::ui::Widget::TouchEventType::BEGAN)
            this->_left = true;
        if (type == cocos2d::ui::Widget::TouchEventType::ENDED)
            this->_left = false;
        
        this->updateMove(-1);
    });
    
    // right
    button = cocos2d::utils::findChild<cocos2d::ui::Button*>(_scene, "right");
    button->addTouchEventListener([=](cocos2d::Ref*, cocos2d::ui::Widget::TouchEventType type){
        if (type == cocos2d::ui::Widget::TouchEventType::BEGAN)
            this->_right = true;
        if (type == cocos2d::ui::Widget::TouchEventType::ENDED)
            this->_right = false;
        
        this->updateMove(-1);
    });
    
    // change weapon L
    button = cocos2d::utils::findChild<cocos2d::ui::Button*>(_scene, "ChangeWeaponL");
    button->addClickEventListener([=](cocos2d::Ref*){
        this->switchWeaponL();
    });
    
    // change weapon R
    button = cocos2d::utils::findChild<cocos2d::ui::Button*>(_scene, "ChangeWeaponR");
    button->addClickEventListener([=](cocos2d::Ref*){
        this->switchWeaponR();
    });
    
    scheduleUpdate();
}

void DragonBonesHandler::update(float dt)
{
    Node::update(dt);
    updatePosition();
    updateAnim();
    updateAttack();
    enterFrameHandler(dt);
}

void DragonBonesHandler::onEnter()
{
    Node::onEnter();
    
    // touch events
    cocos2d::EventListenerTouchAllAtOnce* touchLisener = cocos2d::EventListenerTouchAllAtOnce::create();
    touchLisener->onTouchesBegan = [=](const std::vector<cocos2d::Touch*>&, cocos2d::Event*) -> bool{
        this->attack(true);
        return true;
    };
    touchLisener->onTouchesMoved = [=](const std::vector<cocos2d::Touch*>& touches, cocos2d::Event*){
        auto touchLoc = touches[0]->getLocation();
        this->aim(touchLoc.x, touchLoc.y);
    };
    touchLisener->onTouchesEnded = [=](const std::vector<cocos2d::Touch*>&, cocos2d::Event*){
        this->attack(false);
    };
    cocos2d::Director::getInstance()->getEventDispatcher()->addEventListenerWithSceneGraphPriority(touchLisener, _armatureDisplay);
}

void DragonBonesHandler::onExit()
{
    Node::onExit();
    // since `CCArmatureDisplay` reset EventDispacher, so we should remove event listener manually
    cocos2d::Director::getInstance()->getEventDispatcher()->removeEventListenersForTarget(_armatureDisplay);
    
    dragonBones::CCFactory::destroyInstance();
}

void DragonBonesHandler::animationEventHandler(dragonBones::EventObject* event)
{
    if (event->type == dragonBones::EventObject::FADE_IN_COMPLETE)
    {
        if (event->animationState->getName() == "jump_1")
        {
            _isJumpingB = true;
            _speedY = -DragonBonesHandler::JUMP_SPEED;
            _armature->getAnimation().fadeIn("jump_2", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
        }
        else if (event->animationState->getName() == "jump_4")
            updateAnimation();
    }
    else if (event->type == dragonBones::EventObject::FADE_OUT_COMPLETE)
    {
        if (event->animationState->getName() == "attack_01")
        {
            _isAttackingB = false;
            _attackState = nullptr;
        }
    }
}

void DragonBonesHandler::frameEventHandler(dragonBones::EventObject* event)
{
    if (event->animationState->getName() == "fire_01")
    {
        auto firePointBone = event->armature->getBone("firePoint");
        auto localPoint = cocos2d::Vec2(firePointBone->global.x, -firePointBone->global.y);
        
        auto display = (dragonBones::CCArmatureDisplay*)event->armature->getDisplay();
        auto globalPoint = display->convertToWorldSpace(localPoint);
        
        fire(globalPoint);
    }
}

void DragonBonesHandler::updateAnimation()
{
    if (_isJumpingA)
        return;
    
    if (_isSquating)
    {
        _speedX = 0.f;
        _armature->getAnimation().fadeIn("squat", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
        _walkState = nullptr;
        return;
    }
    
    if (_moveDir == 0)
    {
        _speedX = 0.f;
        _armature->getAnimation().fadeIn("idle", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
        _walkState = nullptr;
    }
    else
    {
        if (!_walkState)
            _walkState = _armature->getAnimation().fadeIn("walk", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
        
        if (_moveDir * _faceDir > 0)
        {
            _walkState->timeScale = DragonBonesHandler::MAX_MOVE_SPEED_FRONT / DragonBonesHandler::NORMALIZE_MOVE_SPEED;
            _speedX = DragonBonesHandler::MAX_MOVE_SPEED_FRONT * _faceDir;
        }
        else
        {
            _walkState->timeScale = -DragonBonesHandler::MAX_MOVE_SPEED_BACK / DragonBonesHandler::NORMALIZE_MOVE_SPEED;
            _speedX = -DragonBonesHandler::MAX_MOVE_SPEED_BACK * _faceDir;
        }
    }
}

void DragonBonesHandler::jump()
{
    if (_isJumpingA)
        return;
    
    _isJumpingA = true;
    _armature->getAnimation().fadeIn("jump_1", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
    _walkState = nullptr;
}

void DragonBonesHandler::squat(bool isSquating)
{
    if (_isSquating == isSquating)
        return;
    
    _isSquating = isSquating;
    updateAnimation();
}

void DragonBonesHandler::updateMove(int dir)
{
    if (_left && _right)
        move(dir);
    else if (_left)
        move(-1);
    else if (_right)
        move(1);
    else
        move(0);
}

void DragonBonesHandler::move(int dir)
{
    if (_moveDir == dir)
        return;
    
    _moveDir = dir;
    updateAnimation();
}

void DragonBonesHandler::aim(float x, float y)
{
    if (_aimDir == 0)
        _aimDir = 10;
    
    _target = _armatureDisplay->getParent()->convertToNodeSpace(cocos2d::Vec2(x, y));
}

void DragonBonesHandler::attack(bool isAttacking)
{
    if (_isAttackingA == isAttacking)
        return;
    
    _isAttackingA = isAttacking;
}

void DragonBonesHandler::updatePosition()
{
    if (_speedX != 0.f)
    {
        auto armatureDisplayPos = _armatureDisplay->getPosition();
        _armatureDisplay->setPositionX(armatureDisplayPos.x + _speedX);
        auto visibleRect = cocos2d::Director::getInstance()->getVisibleSize();
        auto visibleOrigin = cocos2d::Director::getInstance()->getVisibleOrigin();
        auto minX = visibleOrigin.x + 10;
        auto maxX = visibleRect.width + visibleOrigin.x;
        if (armatureDisplayPos.x < minX)
            _armatureDisplay->setPositionX(minX);
        else if (armatureDisplayPos.x > maxX)
            _armatureDisplay->setPositionX(maxX);
    }
    
    if (_speedY != 0)
    {
        if (_speedY > 5 && (_speedY + DragonBonesHandler::G) <= 5)
            _armature->getAnimation().fadeIn("jump_3", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
        
        _speedY += DragonBonesHandler::G;
        
        auto armatureDisplayPos = _armatureDisplay->getPosition();
        _armatureDisplay->setPositionY(armatureDisplayPos.y + _speedY);
        if (armatureDisplayPos.y < DragonBonesHandler::GROUND)
        {
            _armatureDisplay->setPositionY(DragonBonesHandler::GROUND);
            _isJumpingA = false;
            _isJumpingB = false;
            _speedX = 0.f;
            _speedY = 0.f;
            _armature->getAnimation().fadeIn("jump_4", -1, -1, 0, DragonBonesHandler::NORMAL_ANIMATION_GROUP);
            if (_isSquating || _moveDir)
                updateAnimation();
        }
    }
}

void DragonBonesHandler::updateAnim()
{
    if (_aimDir == 0)
        return;
    
    auto position = _armatureDisplay->getPosition();
    _faceDir = _target.x > position.x ? 1 : -1;
    if (_armatureDisplay->getScaleX() * _faceDir < 0)
    {
        _armatureDisplay->setScaleX(_armatureDisplay->getScaleX() * -1);
        if (_moveDir)
            updateAnimation();
    }
    
    auto animOffsetY = _armature->getBone("chest")->global.y * _armatureDisplay->getScaleY();
    if (_faceDir > 0)
        _animRadian = std::atan2(-(_target.y - position.y + animOffsetY), _target.x - position.x);
    else
    {
        _animRadian = M_PI - std::atan2(-(_target.y - position.y + animOffsetY), _target.x - position.x);
        if (_animRadian > M_PI)
            _animRadian -= M_PI * 2;
    }
    
    int aimDir = 0;
    if (_animRadian > 0)
        aimDir = -1;
    else
        aimDir = 1;
    
    if (_aimDir != aimDir)
    {
        _aimDir = aimDir;
        
        if (_aimDir >=0)
            _animState = _armature->getAnimation().fadeIn("aimUp", 0, 1, 0,
                                                          DragonBonesHandler::AIM_ANIMATION_GROUP,
                                                          dragonBones::AnimationFadeOutMode::SameGroup);
        else
            _animState = _armature->getAnimation().fadeIn("aimDown", 0, 1, 0,
                                                          DragonBonesHandler::AIM_ANIMATION_GROUP,
                                                          dragonBones::AnimationFadeOutMode::SameGroup);
    }
    
    _animState->weight = std::fabs(_animRadian / M_PI * 2);
    _armature->invalidUpdate();
}

void DragonBonesHandler::updateAttack()
{
    if (!_isAttackingA || _isAttackingB)
        return;
    
    _isAttackingB = true;
    
    _attackState = _armature->getAnimation().fadeIn("attack_01", -1, -1, 0,
                                                    DragonBonesHandler::ATTACK_ANIMATION_GROUP,
                                                    dragonBones::AnimationFadeOutMode::SameGroup);
    
    _attackState->autoFadeOutTime = _attackState->fadeTotalTime;
    _attackState->addBoneMask("pelvis");
}

void DragonBonesHandler::enterFrameHandler(float dt)
{
    auto iter = _bullets.begin();
    while (iter != _bullets.end())
    {
        if ((*iter)->update())
            iter = _bullets.erase(iter);
        else
            ++iter;
    }
        
    dragonBones::WorldClock::clock.advanceTime(dt);
}

void DragonBonesHandler::switchWeaponL()
{
    static_cast<CCArmatureDisplay*>(_weaponL->_display)->removeEvent(dragonBones::EventObject::FRAME_EVENT);
    
    _weaponLIndex = (_weaponLIndex + 1) % DragonBonesHandler::WEAPON_L_LIST.size();
    auto newWeaponName = DragonBonesHandler::WEAPON_L_LIST[_weaponLIndex];
    _weaponL = dragonBones::CCFactory::getInstance()->buildArmature(newWeaponName);
    _armature->getSlot("weapon_l")->setChildArmature(_weaponL);
    
    static_cast<CCArmatureDisplay*>(_weaponL->_display)->addEvent(dragonBones::EventObject::FRAME_EVENT, CC_CALLBACK_1(DragonBonesHandler::frameEventHandler, this));
}

void DragonBonesHandler::switchWeaponR()
{
    static_cast<CCArmatureDisplay*>(_weaponR->_display)->removeEvent(dragonBones::EventObject::FRAME_EVENT);
    
    _weaponRIndex = (_weaponRIndex + 1) % DragonBonesHandler::WEAPON_R_LIST.size();
    auto newWeaponName = DragonBonesHandler::WEAPON_R_LIST[_weaponRIndex];
    _weaponR = dragonBones::CCFactory::getInstance()->buildArmature(newWeaponName);
    _armature->getSlot("weapon_r")->setChildArmature(_weaponR);
    
    static_cast<CCArmatureDisplay*>(_weaponR->_display)->addEvent(dragonBones::EventObject::FRAME_EVENT, CC_CALLBACK_1(DragonBonesHandler::frameEventHandler, this));
}

void DragonBonesHandler::fire(const cocos2d::Vec2& pos)
{
    auto firePoint = pos;
    firePoint.x += cocos2d::random(0, 1) * 2 - 1;
    firePoint.y += cocos2d::random(0, 1) * 2 - 1;
    
    auto armature = dragonBones::CCFactory::getInstance()->buildArmature("bullet_01");
    auto effect = dragonBones::CCFactory::getInstance()->buildArmature("fireEffect_01");
    auto radian = _faceDir < 0 ? M_PI - _animRadian : _animRadian;
    auto bullet = std::make_shared<DragonBoneBullet>(_parent, armature, effect, radian + cocos2d::random(0, 1) * 0.02f - 0.01f, 40, firePoint);
    addBullet(bullet);
}

void DragonBonesHandler::addBullet(const std::shared_ptr<DragonBoneBullet>& bullet)
{
    _bullets.push_back(bullet);
}

/***************************************
 DragonBoneBullet
 **************************************/
DragonBoneBullet::DragonBoneBullet(cocos2d::Node *parent, dragonBones::Armature *armature, dragonBones::Armature *effect,
                 float radian, float speed, const cocos2d::Vec2& position)
{
    _speedX = std::cos(radian) * speed;
    _speedY = -std::sin(radian) * speed;
    auto thePos = parent->convertToNodeSpace(position);
    
    _armature = armature;
    _armatureDisplay = static_cast<dragonBones::CCArmatureDisplay*>(_armature->_display);
    _armatureDisplay->setPosition(thePos);
    _armatureDisplay->setRotation(radian * dragonBones::RADIAN_TO_ANGLE);
    _armature->getAnimation().play("idle");
    
    if (effect)
    {
        _effect = effect;
        auto effectDisplay = static_cast<dragonBones::CCArmatureDisplay*>(_effect->_display);
        effectDisplay->setRotation(radian * dragonBones::RADIAN_TO_ANGLE);
        effectDisplay->setPosition(thePos);
        effectDisplay->setScaleX(1 + cocos2d::random(0, 1));
        effectDisplay->setScaleY(1 + cocos2d::random(0, 1) * 0.5f);
        if (cocos2d::random() < 0.5f)
            effectDisplay->setScaleY(effectDisplay->getScaleY() * -1);
        
        _effect->getAnimation().play("idle");
        dragonBones::WorldClock::clock.add(_effect);
        parent->addChild(effectDisplay);
    }
    
    dragonBones::WorldClock::clock.add(_armature);
    parent->addChild(_armatureDisplay);
}

bool DragonBoneBullet::update()
{
    _armatureDisplay->setPosition(_armatureDisplay->getPosition() + cocos2d::Vec2(_speedX, _speedY));
    
    auto worldPos = _armatureDisplay->getParent()->convertToWorldSpace(_armatureDisplay->getPosition());
    auto visibleRect = cocos2d::Director::getInstance()->getVisibleSize();
    auto visibleOrigin = cocos2d::Director::getInstance()->getVisibleOrigin();
    if (worldPos.x < visibleOrigin.x || worldPos.x >= visibleOrigin.x + visibleRect.width ||
        worldPos.y < visibleOrigin.y || worldPos.y >= visibleOrigin.y + visibleRect.height)
    {
        doClean();
        return true;
    }
    
    return false;
}

void DragonBoneBullet::doClean()
{
    dragonBones::WorldClock::clock.remove(_armature);
    _armatureDisplay->removeFromParent();
    _armature->dispose();
    
    if (_effect)
    {
        dragonBones::WorldClock::clock.remove(_effect);
        static_cast<CCArmatureDisplay*>(_effect->_display)->removeFromParent();
        _effect->dispose();
    }
}
